pkg/vex/vex.go

```diff
@@ -48,6 +182,121 @@ func FromPackageConfiguration(vexCfg Config, buildCfg ...*build.Configuration) (
 	return doc, nil
 }
 
+// indexMelangeConfigsDir reads the distro config directory and
+// returns an indexed map where keys are packages and values are
+// its connfiguration file.
+func indexMelangeConfigsDir(dirPath string) (map[string]string, error) {
+	finfo, err := os.Stat(dirPath)
+	if err != nil {
+		return nil, fmt.Errorf("checking the config directory")
+	}
+	if !finfo.IsDir() {
+		return nil, errors.New("distro config path is not a directory")
+	}
+
+	files, err := filepath.Glob(fmt.Sprintf("%s/*.yaml", dirPath))
+	if err != nil {
+		return nil, fmt.Errorf("listing configuration files")
+	}
+
+	confMap := map[string]string{}
+	for _, f := range files {
+		buildCfg, err := build.ParseConfiguration(f)
+		if err != nil {
+			return nil, fmt.Errorf("parsing %s config: %w", f, err)
+		}
+		confMap[buildCfg.Package.Name] = f
+
+		for i := range buildCfg.Subpackages {
+			confMap[buildCfg.Subpackages[i].Name] = f
+		}
+	}
+	return confMap, nil
+}
+
+// extractPackagePurls returns all purls describing distro apks found
+// in elements related to a SPDX package
+//
+//nolint:gocritic
+func extractPackagePurls(vexCfg Config, sbom *spdx.Document, spdxID string, seen *map[string]struct{}) []purl.PackageURL {
+	purls := []purl.PackageURL{}
+	for _, r := range sbom.Relationships {
+		if r.Element != spdxID {
+			continue
+		}
+		if _, ok := (*seen)[r.Related]; ok {
+			continue
+		}
+		(*seen)[r.Related] = struct{}{}
+		subpurls := extractPackagePurls(vexCfg, sbom, r.Related, seen)
+		purls = append(purls, subpurls...)
+		for _, subpackage := range sbom.Packages {
+			if subpackage.ID != r.Related {
+				continue
+			}
+			for _, ref := range subpackage.ExternalRefs {
+				if ref.Type != "purl" {
+					continue
+				}
+				// If malformed, just skip it
+				p, err := purl.FromString(ref.Locator)
+				if err != nil {
+					continue
+				}
+
+				if p.Namespace == vexCfg.Distro {
+					purls = append(purls, p)
+				}
+			}
+		}
+	}
+	return purls
+}
+
+// extractSBOMPurls reads an SBOM and returns the purls identifying
+// packages from the distribution.
+func extractSBOMPurls(vexCfg Config, sbom *spdx.Document) map[string][]purl.PackageURL {
+	purls := map[string][]purl.PackageURL{}
+	for _, elementID := range sbom.DocumentDescribes {
+		packagePurls := extractPackagePurls(vexCfg, sbom, elementID, &map[string]struct{}{})
+		if len(packagePurls) == 0 {
+			continue
+		}
+
+		descPurl := ""
+		for i := range sbom.Packages {
+			if sbom.Packages[i].ID != elementID {
+				continue
+			}
+			for _, ref := range sbom.Packages[i].ExternalRefs {
+				if ref.Type == "purl" {
+					descPurl = ref.Locator
+				}
+			}
```

Comment:
Just so I understand, does this mean "last purl wins"? I'm guessing that's fine

Reply:
Yes, although we filter purls to only those namespaced to Wolfi (the distro param) so it should always be just one. Perhaps we can spare the cycle there.